### DISCUSSION TITLE
snap-proposal: Teaches `pct` options new tricks so you don't need to use `mm` options

### DISCUSSION THREAD

**Discussion by joostdecock** [freesewing/freesewing#1331] [Support] [answered]
Fresh on the heels of my design that [Patterns with `mm` options will no longer be accepted in our repository](https://github.com/freesewing/freesewing/discussions/1326) here is something I've cherry-picked from [the v3 roadmap](https://github.com/freesewing/freesewing/discussions/1278) and ported to v2 (aka, the current version) because it could be accomplished without breaking backwards compatibility.

I am writing this up as a proposal because even though it's a (potential) v2 update, it is very much in the spirit of where we want to go with v3, and so I welcome input feedback from the community.

So let's have a look at what it's all about.

## What's wrong with `mm` options?

They don't scale. Parametric design is the *raison d'être* of FreeSewing and that core belief that things should seamlessly adapt goes out the window when you use a `mm` option because now you have a value that will not change based on the input measurements.

There are some who  argue that *it's fine because you can just lower the option* but that breaks the principle of **sensible defaults** (aka [no surprises](https://www.youtube.com/watch?v=u5CVsCnxyXg)). The fact that you can sidestep the bullet does not mean you're not creating a *footgun*.

## But reasons!

People love `mm` options, and have good reasons for that. Wouter, one of the elders of FreeSewing, summarizes the objections better than I can:

> I'm not sure that is possible in real life. Elastics only come in certain widths. A bow tie band (with holes for adjusting the size of the bow tie) is only available in one width, and with specific spacing between the holes. Snaps and other hardware only come in certain sizes, and not in a percentage of anything.

In our years working together I have learned that when Wouter says something, it is wise to pay attention. 
And he and others who protested have a good point. These are valid concerns that deserve a better answer than the somewhat vague promise that *we'll fix it in v3, look it's already on the roadmap*.

So, after mulling it over for a while, and looking into how such a thing would work, I came up with this proposal which implements/backports 2 things from the v3 roadmap into the current codebase:

 - Support for reporting the absolute value of options
 - Support snapping of percentage options

To be clear, I really just set out ot fix the second item, the so-called **snapping of percentage options**. It just so happens that when thinking about how this would work and how to implement it, I also did the groundwork for the first item. However, that will also require a bunch of work in the frontend, so the remainder of this proposal I will focus (almost) exclusively on the snapping of percentage options.

## What does this proposal implement?

It implements **snapping** of percentage options to **absolute values** (think mm). To make that happen, it does the following:

 - Calculate the **absolute value** of the option
 - See whether that approaches one of the absolute values to snap to (more on what 'approaching' means later)
 - If so, set the absolute option to the snap
 - If not, set the absolute option to whatever the absolute option is

If you're head's spinning, here's an image that will hopefully clarify things a bit:

![snap-proposal](https://user-images.githubusercontent.com/1708494/133493652-9a23a24d-ba68-4280-8e95-1ca38cf418bc.png)

The gradient box represents the range of any given measurement, from dolls all the way on the left, to giants all the way on the right.
The sort of middle green-colored region is what the designer had in mind when designing the pattern, and they have setup snap points for values that they feel make sense.

The region of influence of any given snap point will extend 50% towards it's neighbor, on both sides (indicated by the dashed lines).This means that the region of snap points is continuous, once you're in, you're going to be snapped to a given absolute value.

However, when you venture out into the area where the designer did not configure any snap points, your input will be used as-is.

This system results in a best of both words:

 - Things like elastic widths and so on can be configured to be fixed values, of common elastic widths for example
 - The default value will still scale up and down, but merely snap to the closest fixed value
 - When the input measurements go somewhere the designer did not anticipate, the option will just behave as a regular `pct` option

I feel this is a solution that addresses the criticism of Wouter and others, without jeopardizing the parametric nature of our patterns.

## What does it take?

All this comes at a price of course. We need to do two things:

 - Define what values to snap to
 - Provide a way to determine what the **absolute value** of a `pct` option will be

### Define what values to snap to

There are three different scenarios:

#### snap is a number

In this case, the **absolute value** of the option will be *snapped* to a multiple of this value. An example:

```js
myOption: {
  pct:5,
  min: 0
  max: 35,
  snap: 7
}
```

The absolute value of this option will be set to a multiple of `7` (so one of 7, 14, 21, 28, 35, ...)

Note that in a case like this, the value will **always** be snapped, because the snap points will be distributed equally across the entire range of all possible input.

#### snap is an array of numbers

This will snap the **absolute value** of the option to any of the numbers. An example:

```js
myOption: {
  pct:5,
  min: 0
  max: 35,
  snap: [7, 12, 21, 34, 53, 64 ]
}
```

This will snap the **absolute value** of the option to the number of the list, if it approaches it. (see below for what that means).

#### snap is an object with `metric` and `imperial` properties that each hold an array of numbers

This will work the same as a list of snaps, but it allows you to supply a different list for users drawing metric, or imperial patterns. An example config:

```js
myOption: {
  pct:5,
  min: 0
  max: 35,
  snap: {
    metric: [7, 12, 21, 34, 53, 64 ],
    imperial: [25.4, 50.8, 76.3 ],
  }
}
```

This will snap the **absolute value** of the option to the number of the list, if it approaches it. (as explained above).

### Provide a way to determine what the absolute value of a `pct` option will be

For this, `pct` options take the `toAbs` propertiy that is a function with the following signature:

```js
function toAbs(pctValue, settings) {
  // return absolute value here
}
```
The first parameter is the value of the option provided by the user, the second is the pattern's settings object which (among other things) holds the measurements.

In the most common scenario, your option is a simple fraction of a measurement (head circumference in the example below), so with object destructuring and fat-arrow notation, you can write it a bit terser like this:

```js
toAbs: (val, { measurement }) => measurements.head * val
```

As this scenario is so common, `@freesewing/core` now exports a `pctBasedOn` method that you can spread into your option config.
Here's an example taken from the paco config file in the `snap-proposal` feature branch:

First import the method:
https://github.com/freesewing/freesewing/blob/025cf9b88e2de9d35a084ed62d6fa3232e0fef10/packages/paco/config/index.js#L2

Then use it in your option with the name of the measurement the option depends on:

https://github.com/freesewing/freesewing/blob/025cf9b88e2de9d35a084ed62d6fa3232e0fef10/packages/paco/config/index.js#L100-L107

This will not only add an `toAbs()` method to your option that calculates the absolute value based on the measurement you specify. It will also add a `fromAbs()` method that does the inverse: return the pct value required to achieve a certain absolute value, which is another item on our v3 roadmap.

## How to use this in your pattern code

This is all good and well, but how do you use this?

Well, just like you can get the `options` object from our shorthand call, you can now get the `absoluteOptions` object that holds absolute values for those options with snaps configured.

In our paco example, what used to be:

```js
store.set('ankleElastic', measurements.waistToFloor * options.ankleElastic)
```

is now:

```js
store.set('ankleElastic', absoluteOptions.ankleElastic)
```

> **Note**: There's really no added value in setting this in the store as absoluteOptions is available everywhere, but I wanted to change as little as possble

## The end result

Here's a screenshot of Paco's `ankleElastic` option sampled. As you can see, it's not uniformly distributed because of the snapping taking effect:

![paco](https://user-images.githubusercontent.com/1708494/133497758-b705c47c-316b-4aa9-9709-8aaa0d51c92f.png)

## When to use this in your pattern code

Don't go rushing out turning every `pct` option into a snapped percent option. This makes sense in scenarios where you would be tempted to reach for the `mm` options. Elastic width, waistband width, button size, other things that come in certain sizes, you get the idea.

## The code

This proposal is implemented in [this commit](https://github.com/freesewing/freesewing/commit/025cf9b88e2de9d35a084ed62d6fa3232e0fef10). You can kick the tires by checking out [the `snap-proposal` feature branch](https://github.com/freesewing/freesewing/tree/snap-proposal).

## Feedback / Rants / Raves

I propose to add this to the next  minor release of FreeSewing. But not before collecting input from others since this is a significant new feature, and I want to make sure that people are happy with the developer experience.

So please don't be shy and share your thoughts below.

joost

**Comment by woutervdub** [comment]
I like this a lot. I'd have to play with it in a pattern to see how it works. The idea is solid as far as I can tell. When I have some time (not for the next 3 weeks, as my daughter is getting married), I'll put it in Hortensia, since that one has to deal with a zipper. That is now implemented as individual choices, but could easily be related to the whole sizing of the bag.

**Comment by joostdecock** [comment, marked as answer]
This has been implemented a while back

**Comment by joostdecock** [comment]
Done. Closing
